### DISCUSSION TITLE
Support LINQ Where/First operators in GridifyMapper

### DISCUSSION THREAD

**Discussion by uxohotteixo** [alirezanet/Gridify#158] [General]
Hi @alirezanet!

In addition to using .Select and .SelecyMany for nested collections (which works really fine) I've tried the following expressions:

First:
  ```
 var query = new QueryBuilder<Entity>()
                    .UseEmptyMapper()
                    .AddMap("model", t => t.SubCollection.Where(sc => sc.Prop == "1").Select(s => s.Entity2.Name))
                    .AddCondition("model=someFilterValue")
                    .BuildFilteringExpression();
```

Second:
  ```
 var query = new QueryBuilder<Entity>()
                    .UseEmptyMapper()
                    .AddMap("model", t => t.SubCollection
                        .Where(sb => sb.State == States.InUse)
                        .OrderBy(s => s.Order)
                        .Select(s => s.Entity1.Entity2.Description)
                        .First())
                    .AddCondition("model=someFilterValue")
                    .BuildFilteringExpression();
```
In Linq such expressions work as expected but when I try to insert this into GridifyMapper I get various internal exceptions. I can provide those as well if required.

**Comment by uxohotteixo** [comment]
Now I know the only working approach here is to make a projection (for subcollections too) then filter and sort the projection in Gridify. But projection before filtering in EF seems to be a bad solution and will affect the quality of queries

**Reply by alirezanet** [reply]
Hi @uxohotteixo,
 In the `AddMap` method, you can't use any filtering or `Where` clause since it just maps a string to the related properties. otherwise, mapping doesn't make sense. 
but you can use the mapped properties in the `AddConidtion` method to set a default filtering on your data. 
so basically according to your first example, you can map `sc.Prop` to something like `AddMap("myProp" ,  ...)` and use it in AddCondition method like `.AddCondition("myprop=1")`

**Reply by alirezanet** [reply]
The second way that comes to my mind in this scenario is to add your hard query to your collection and then use Gridify to filter it out. 
```
var query = queriableCollection.Where(sc => sc.Any(x.Prop == "1")).ApplyFiltering(gridifyQuery);
```

**Reply by uxohotteixo** [reply]
@alirezanet Thank you! So I decided to move the interfering hard LINQ in the AutoMapper projection and use GridifyMapper as it should be! :)
